### PR TITLE
feat(react-grid): allow setting min and max width limitation for every column

### PR DESCRIPTION
**Pull request by LazyLahtak**
Feat #995

### CODE REVIEW

File: packages/dx-grid-core/src/plugins/table-column-resizing/reducers.ts [hunk 1]
```diff
@@ -17,11 +32,26 @@ export const changeTableColumnWidth: ColumnWidthReducer = (
 };
 
 export const draftTableColumnWidth: ColumnWidthReducer = (
-  state, { columnName, shift, minColumnWidth },
+  state, { columnName, shift, minColumnWidth, maxColumnWidth, columnExtensions },
 ) => {
   const { columnWidths } = state;
   const updatedColumn = columnWidths.find(elem => elem.columnName === columnName)!;
-  const size = Math.max(minColumnWidth, updatedColumn.width + shift);
+  const indexOfExtended = columnExtensions
```
Comment (ushkal) [inline packages/dx-grid-core/src/plugins/table-column-resizing/reducers.ts]: `changeTableColumnWidth` and `draftTableColumnWidth` should be refactored, repeated code should be extracted

**Comment by LazyLahtak** [comment]
FPR

File: packages/dx-grid-core/src/plugins/table-column-resizing/reducers.ts [hunk 2]
```diff
@@ -1,28 +1,52 @@
 import { slice } from '@devexpress/dx-core';
-import { ColumnWidthReducer } from '../../types';
+import { ColumnWidthReducer, ColumnSizeFn } from '../../types';
+
+export const getColumnSize: ColumnSizeFn = (
+  updatedColumn, {columnName, shift, minColumnWidth, maxColumnWidth, columnExtensions,
+}) => {
+  const indexOfExtended = columnExtensions
```
Comment (ushkal) [inline packages/dx-grid-core/src/plugins/table-column-resizing/reducers.ts]: Specify the `columnExtensions` default value, an empty array. Then ternary operator can be removed

Comment (ushkal) [reply packages/dx-grid-core/src/plugins/table-column-resizing/reducers.ts]: And we can use `find` here

File: packages/dx-grid-core/src/types/column-resizing.types.ts [hunk 3]
```diff
@@ -9,6 +9,14 @@ export interface TableColumnWidthInfo {
   /** A column width. */
   width: number;
 }
+export interface ColumnExtension {
+  /** A column name. */
+  columnName: string;
+  /** A column min width. */
```
Comment (ushkal) [inline packages/dx-grid-core/src/types/column-resizing.types.ts]: min, max -> minimum, maximum

File: packages/dx-grid-core/src/types/column-resizing.types.ts [hunk 4]
```diff
@@ -9,6 +9,14 @@ export interface TableColumnWidthInfo {
   /** A column width. */
   width: number;
 }
+export interface ColumnExtension {
```
Comment (ushkal) [inline packages/dx-grid-core/src/types/column-resizing.types.ts]: Hide it in a namespace like column extensions in other plugins

**Comment by ushkal** [comment]
The docs should be done too

File: packages/dx-grid-core/src/types/column-resizing.types.ts [hunk 5]
```diff
@@ -9,6 +9,14 @@ export interface TableColumnWidthInfo {
...
+  /** A column name. */
+  columnName: string;
+  /** A column min width. */
+  minWidth: number;
+  /** A column max width. */
+  maxWidth: number;
```
Comment (MaximKudriavtsev) [inline packages/dx-grid-core/src/types/column-resizing.types.ts]: Are these properties required?

**Comment by LazyLahtak** [comment]
FPR

File: packages/dx-react-grid/src/types/column-operations/column-resizing.types.ts [hunk 6]
```diff
@@ -1,10 +1,25 @@
 import { TableColumnWidthInfo } from '../index';
-
+// tslint:disable-next-line:no-namespace
+export namespace TableColumnResizing {
```
Comment (ushkal) [inline packages/dx-react-grid/src/types/column-operations/column-resizing.types.ts]: This type is declared in grid-core, it should be re-exported from there

File: packages/dx-grid-core/src/plugins/table-column-resizing/reducers.ts [hunk 7]
```diff
@@ -1,28 +1,50 @@
 import { slice } from '@devexpress/dx-core';
-import { ColumnWidthReducer } from '../../types';
+import { ColumnWidthReducer, ColumnSizeFn } from '../../types';
+
+export const getColumnSize: ColumnSizeFn = (
+  updatedColumn, {columnName, shift, minColumnWidth, maxColumnWidth, columnExtensions = [],
+}) => {
+  const indexOfExtended = columnExtensions.findIndex(elem => elem.columnName === columnName);
```
Comment (ushkal) [inline packages/dx-grid-core/src/plugins/table-column-resizing/reducers.ts]: We should rather use `find` here

File: packages/dx-grid-core/src/plugins/table-column-resizing/reducers.ts [hunk 8]
```diff
@@ -1,28 +1,50 @@
...
+  const extendedColumn = indexOfExtended >= 0 ? columnExtensions[indexOfExtended] : undefined;
+  const haveExtendedMinWidth = extendedColumn && extendedColumn.minWidth! >= 0;
+  const haveExtendedMaxWidth = extendedColumn && extendedColumn.maxWidth! >= 0;
```
Comment (ushkal) [inline packages/dx-grid-core/src/plugins/table-column-resizing/reducers.ts]: This condition is trivial and can be used in ternary as-is

File: packages/dx-grid-core/src/plugins/table-column-resizing/reducers.test.ts [hunk 9]
```diff
@@ -6,64 +6,201 @@ import {
...
+        minColumnWidth: 40,
+        maxColumnWidth: 80,
+        columnExtensions,
+      })).
+        toEqual({
+          columnWidths: [{ columnName: 'a', width: 40 }, { columnName: 'b', width: 70 }],
+        });
+    });
+
+    it('should work if limitation does not define in columnExtensions', () => {
```
Comment (ushkal) [inline packages/dx-grid-core/src/plugins/table-column-resizing/reducers.test.ts]: -> is not defined

File: packages/dx-react-grid/src/plugins/table-column-resizing.test.tsx [hunk 10]
```diff
@@ -145,4 +151,34 @@ describe('TableColumnResizing', () => {
...
+        <TableColumnResizing
+          {...defaultProps}
+          defaultColumnWidths={[{ columnName: 'a', width: 100 }]}
+          columnExtensions={columnExtensions}
+        />
+      </PluginHost>
+    ));
+
+    draftTableColumnWidth.mockReturnValue({ draftColumnWidths: [{ columnName: 'a', width: 150 }] });
+    executeComputedAction(tree, actions => actions.draftTableColumnWidth(payload));
```
Comment (ushkal) [inline packages/dx-react-grid/src/plugins/table-column-resizing.test.tsx]: We should test `changeTableColumnWidth` too (in a separate test)

File: packages/dx-react-grid/src/types/column-operations/column-resizing.types.ts [hunk 11]
```diff
@@ -1,10 +1,25 @@
-import { TableColumnWidthInfo } from '../index';
-
+import { TableColumnWidthInfo } from '@devexpress/dx-grid-core';
```
Comment (ushkal) [inline packages/dx-react-grid/src/types/column-operations/column-resizing.types.ts]: It should be re-exported in index like other types

File: packages/dx-grid-core/src/plugins/table-column-resizing/reducers.ts [hunk 12]
```diff
@@ -1,28 +1,47 @@
 import { slice } from '@devexpress/dx-core';
-import { ColumnWidthReducer } from '../../types';
+import { ColumnWidthReducer, ColumnSizeFn } from '../../types';
+
+export const getColumnSize: ColumnSizeFn = (
+  updatedColumn, {columnName, shift, minColumnWidth, maxColumnWidth, columnExtensions = [],
+}) => {
+  const extendedColumn = columnExtensions.find(elem => elem.columnName === columnName);
+  const minWidth = extendedColumn && extendedColumn.minWidth
```
Comment (ushkal) [inline packages/dx-grid-core/src/plugins/table-column-resizing/reducers.ts]: We should check if width >= 0

**Comment by LazyLahtak** [comment]
FPR

**Pull request merged**